runtime/tests/checker/member_test.go

```diff
@@ -418,3 +418,246 @@ func TestCheckMemberNotDeclaredSecondaryError(t *testing.T) {
+                init() {
+                    self.x = []
+                }
+            }
+
+            fun test() {
+                let test = Test()
+                var x: [Int] = test.x
+            }
+        `)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("composite, function", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+            struct Test {
+                pub fun foo(): Int {
+                    return 1
+                }
+            }
+
+            fun test() {
+                let test = Test()
+                var foo: (fun(): Int) = test.foo
+            }
+        `)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("composite reference, field", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+            struct Test {
+                var x: [Int]
+                init() {
+                    self.x = []
+                }
+            }
+
+            fun test() {
+                let test = Test()
+                let testRef = &test as &Test
+                var x: &[Int] = testRef.x
+            }
+        `)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("composite reference, optional field", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+            struct Test {
+                var x: [Int]?
+                init() {
+                    self.x = []
+                }
+            }
+
+            fun test() {
+                let test = Test()
+                let testRef = &test as &Test
+                var x: &[Int]? = testRef.x
+            }
+        `)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("composite reference, primitive field", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+            struct Test {
+                var x: Int
+                init() {
+                    self.x = 1
+                }
+            }
+
+            fun test() {
+                let test = Test()
+                let testRef = &test as &Test
+                var x: Int = testRef.x
+            }
+        `)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("composite reference, function", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+            struct Test {
+                pub fun foo(): Int {
+                    return 1
+                }
+            }
+
+            fun test() {
+                let test = Test()
+                let testRef = &test as &Test
+                var foo: (fun(): Int) = testRef.foo
+            }
+        `)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("array, element", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+            fun test() {
+                let array: [[Int]] = [[1, 2]]
+                var x: [Int] = array[0]
+            }
+        `)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("array reference, element", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+            fun test() {
+                let array: [[Int]] = [[1, 2]]
+                let arrayRef = &array as &[[Int]]
+                var x: &[Int] = arrayRef[0]
```

Comment:
Is this the behavior we want? I can see arguments both for and against, but it is not immediately clear from the FLIP (imo) that things should work this way, since I'm not sure it's necessarily the case that array/dictionary elements are really "fields".

Reply:
Yup, this is in accordance with the objective of the FLIP. The FLIP actually explicitly states about the array/dictionary behavior.
Maybe I can add more examples there, if that helps eliminate any potential confusion.

Reply:
This came up in a follow PR as well, and posted a sample code there on why we would need this: https://github.com/onflow/cadence/pull/2587#discussion_r1237231577.

Like it was discussed there, we will return a reference for array/dictionary elements, but it will be an unauthorized one.